File: simple/rest.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"io/ioutil"
	"net/http"
	"io"
	"strconv"
	"time"
	"sync"
	"strings"
)

const (
	get     = "GET"
	post    = "POST"
	website = "https://api.gdax.com"
)

var (
	limit = &sync.Mutex{}
	rate = time.Millisecond * time.Duration(500)
)

func request(method, url string, body string) (*http.Client, *http.Request, error) {
	var post io.Reader
	if body != "" {
		post = strings.NewReader(body)
	}
	request, err := http.NewRequest(method, url, post)
	if err != nil {
		return nil, nil, err
	}
	request.Header.Add("Accept", "application/json")
	request.Header.Add("Content-Type", "application/json")
	request.Header.Add("User-Agent", "napa")
	return &http.Client{}, request, nil
}

func publicRequest(method, path string) ([]byte, int, error) {
	client, request, err := request(method, website+path, "")
	if err != nil {
		return nil, 0, err
	}
	limit.Lock()
	response, err := client.Do(request)
	time.Sleep(rate)
	limit.Unlock()
	if err != nil {
		return nil, 0, err
	}
	defer response.Body.Close()
	read, err := ioutil.ReadAll(response.Body)
	return read, response.StatusCode, err
}

func privateRequest(auth map[string]string, method, path, body string) ([]byte, int, error) {
	client, request, err := request(method, website+path, body)
	if err != nil {
		return nil, 0, err
	}
	limit.Lock()
	timestamp := strconv.FormatInt(time.Now().Unix(), 10)
	what := timestamp + method + path + body
	base64key, err := base64.StdEncoding.DecodeString(auth["secret"])
	if err != nil {
		return nil, 0, err
	}
	hashMessage := hmac.New(sha256.New, base64key)
	_, err = hashMessage.Write([]byte(what))
	if err != nil {
		return nil, 0, err
	}
	signature := base64.StdEncoding.EncodeToString(hashMessage.Sum(nil))
	request.Header.Add("CB-ACCESS-KEY", auth["key"])
	request.Header.Add("CB-ACCESS-SIGN", signature)
	request.Header.Add("CB-ACCESS-TIMESTAMP", timestamp)
	request.Header.Add("CB-ACCESS-PASSPHRASE", auth["phrase"])
	response, err := client.Do(request)
	time.Sleep(rate)
	limit.Unlock()
	if err != nil {
		return nil, 0, err
	}
	defer response.Body.Close()
	read, err := ioutil.ReadAll(response.Body)
	return read, response.StatusCode, err
}

File: python/rest.py
import time
import http.client
import hmac
import hashlib
import time
import base64
import json


def prepare_request(method, site, path, body):
    con = http.client.HTTPSConnection(site, 443)
    if body:
        con.putrequest(method, path, body)
    else:
        con.putrequest(method, path)
    con.putheader('Accept', 'application/json')
    con.putheader('Content-Type', 'application/json')
    con.putheader('User-Agent', 'napa')
    return con


def request(method, site, path, body):
    con = prepare_request(method, site, path, body)
    con.endheaders()
    response = con.getresponse()
    raw_js = response.read()
    status = response.status
    con.close()
    time.sleep(0.5)
    try:
        return json.loads(raw_js.decode()), status
    except Exception:
        return raw_js, status

def private_request(auth, method, site, path, body):
    con = prepare_request(method, site, path, body)
    timestamp = str(time.time())
    message = (timestamp + method + path + body).encode()
    hmac_key = base64.b64decode(auth.secret)
    signature = hmac.new(hmac_key, message, hashlib.sha256)
    signature_b64 = base64.b64encode(signature.digest()).decode()
    con.putheader('CB-ACCESS-KEY', auth.key)
    con.putheader('CB-ACCESS-SIGN', signature_b64)
    con.putheader('CB-ACCESS-TIMESTAMP', timestamp)
    con.putheader('CB-ACCESS-PASSPHRASE', auth.phrase)
    con.endheaders()
    response = con.getresponse()
    raw_js = response.read()
    status = response.status
    con.close()
    time.sleep(0.5)
    try:
        return json.loads(raw_js.decode()), status
    except Exception:
        return raw_js, status
    
    
    